### PR TITLE
PIM-5935: Add show job tracker button on last operations widget

### PR DESCRIPTION
**Pull request by willy-ahva**
last_operations

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | n
| Added Behats                      | y
| Changelog updated                 | n
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

### CODE REVIEW

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 1]
```diff
@@ -5,18 +5,13 @@ define(
...
-        'text!pim/dashboard/template/last-operations-widget'
+        'text!pim/dashboard/template/last-operations-widget',
+        'text!pim/dashboard/template/job-tracker-button'
     ],
-    function ($, _, Routing, Navigation, AbstractWidget, template) {
+    function ($, _, Routing, Navigation, AbstractWidget, template, jobTrackerBtnTemplate) {
         'use strict';
 
         return AbstractWidget.extend({
-            tagName: 'table',
```
Comment (damien-carcel) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: was not used?

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 2]
```diff
@@ -33,11 +28,17 @@ define(
...
             events: {
-                'click a.btn': 'followLink',
-                'click a#btn-show-list': 'showList'
+                'click a.btn': 'followLink'
             },
 
+            /**
+             * Redirect to the clicked operation page
+             *
+             * @param {Object} e
```
Comment (juliensnz) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: event

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 3]
```diff
@@ -62,12 +63,48 @@ define(
...
             },
 
-            showList: function (e) {
+            /**
+             * Add the button which show the job tracker
+             *
+             * @returns {exports.setShowTrackerBtn}
+             */
+            setShowTrackerBtn: function () {
+                var $oldTrackerBtn = $('.show-tracker-btn');
```
Comment (juliensnz) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: this.$(

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 4]
```diff
@@ -62,12 +63,48 @@ define(
                 Navigation.getInstance().setLocation(route);
             },
 
-            showList: function (e) {
+            /**
+             * Add the button which show the job tracker
+             *
+             * @returns {exports.setShowTrackerBtn}
+             */
+            setShowTrackerBtn: function () {
```
Comment (juliensnz) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: why you don't just add it to the main template?

File: features/dashboard/display_last_operations_widget.feature [hunk 5]
```diff
@@ -36,3 +36,10 @@ Feature: Display last operations widget
     Then I should see the "Refresh" button
     And I should see the "Reset" button
     And I should see the columns Type, Job, User, Status and Started at
+
+  Scenario: Show job tracker
+    Given a "footwear" catalog configuration
+    And I am logged in as "Mary"
+    When I am on the dashboard page
+    And I follow "Show job tracker"
+    Then I should see the text "Process tracker"
```
Comment (aRn0D) [inline features/dashboard/display_last_operations_widget.feature]: `Then I should be on the page ...`, we should check if we are on the right page

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 2 above)
Comment (aRn0D) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: we can rename e in event too

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: The last operation template is about the table in the widget. The header is not accessible. It is in the base twig used for all widgets.

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 1 above)
Comment (willy-ahva) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: nope

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 6]
```diff
@@ -33,41 +28,83 @@ define(
 
             template: _.template(template),
 
+            jobTrackerBtnTemplate: _.template(jobTrackerBtnTemplate),
+
             events: {
-                'click a.btn': 'followLink',
-                'click a#btn-show-list': 'showList'
+                'click a.btn': 'followLink'
```
Comment (grena) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Isn't it 
```
'.show-tracker-btn': 'followLink'
```
As you have it in your template?

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 7]
```diff
@@ -33,41 +28,83 @@ define(
...
+                if (0 !== $oldTrackerBtn.length) {
+                    $oldTrackerBtn.remove();
+                }
+
+                this.$jobTrackerBtn = $(this.jobTrackerBtnTemplate());
+
+                this.$el.parent().siblings('.widget-header').append(this.$jobTrackerBtn);
+                this.$jobTrackerBtn.on('click', _.bind(this.showTracker, this));
+
+                return this;
```
Comment (grena) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: you can remove this return statement, and the `@returns` in the jsdoc too.

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 8]
```diff
@@ -33,41 +28,83 @@ define(
...
+            },
+
+            /**
+             * Call when user click on the show job tracker button. Redirect to the Job tracker.
```
Comment (grena) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: click**s**

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 4 above)
Comment (grena) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Imho it's a bad naming, it's not a `setter`
Could be `addShowTrackerBtn`

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 9]
```diff
@@ -33,41 +28,83 @@ define(
...
+            /**
+             * Add the button which show the job tracker
+             *
+             * @returns {exports.setShowTrackerBtn}
+             */
+            setShowTrackerBtn: function () {
+                var $oldTrackerBtn = this.$('.show-tracker-btn');
+
+                if (0 !== $oldTrackerBtn.length) {
+                    $oldTrackerBtn.remove();
```
Comment (grena) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: it's a kind of tricky part, maybe you could add a comment on why you do this?

**Comment by grena** [comment]
GTM once fixed.

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 6 above)
Comment (willy-ahva) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Nope this one is for each lines to follow the link to the operation

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 1 above)
Comment (phaseinducer) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: It's a backbone special property: http://backbonejs.org/#View-el
And it's not defined in the parent so don't remove it.

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 10]
```diff
@@ -5,18 +5,13 @@ define(
...
-    function ($, _, Routing, Navigation, AbstractWidget, template) {
+    function ($, _, Routing, Navigation, AbstractWidget, template, jobTrackerBtnTemplate) {
         'use strict';
 
         return AbstractWidget.extend({
-            tagName: 'table',
-
-            id: 'last-operations-widget',
-
-            className: 'table table-condensed table-light groups unspaced',
```
Comment (phaseinducer) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Are you sure you can remove this?

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 6 above)
Comment (phaseinducer) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Please don't rely on tag names or bootstrap classes to bind your events. Add your own semantic classes if needed.

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js (hunk 10 above)
Comment (willy-ahva) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: yes

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 11]
```diff
@@ -33,41 +28,80 @@ define(
...
+             */
+            addShowTrackerBtn: function () {
+                var $oldTrackerBtn = this.$('.show-tracker-btn');
+
+                // If the button already exists we remove it. It happens because rendering is done several times.
+                if (0 !== $oldTrackerBtn.length) {
+                    $oldTrackerBtn.remove();
+                }
+
+                this.$jobTrackerBtn = $(this.jobTrackerBtnTemplate());
```
Comment (phaseinducer) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Why make it a class property?

File: src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js [hunk 12]
```diff
@@ -33,41 +28,80 @@ define(
...
+
+                this.$el.parent().siblings('.widget-header').append(this.$jobTrackerBtn);
```
Comment (phaseinducer) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: You shouldn't touch anything outside the view. If you need to then there is a bad design somewhere.

Comment (willy-ahva) [inline src/Pim/Bundle/DashboardBundle/Resources/public/js/last-operations-widget.js]: Seen irl. We let it like that.

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**